kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/optimizations/KonanBCEForLoopBodyTransformer.kt

```diff
@@ -49,14 +50,29 @@ class KonanBCEForLoopBodyTransformer : ForLoopBodyTransformer() {
             irExpression.transformChildrenVoid(this)
     }
 
+    private fun IrGetValue.compareConstValue(compare: (IrExpression) -> Boolean): Boolean {
+        val variable = symbol.owner
+        return if (variable is IrVariable && !variable.isVar && variable.initializer != null) {
+            compare(variable.initializer!!)
+        } else false
+    }
+
     private fun IrExpression.compareIntegerNumericConst(compare: (Long) -> Boolean): Boolean {
         @Suppress("UNCHECKED_CAST")
-        return this is IrConst<*> && value is Number && compare((value as Number).toLong())
+        return when (this) {
+            is IrConst<*> -> value is Number && compare((value as Number).toLong())
+            is IrGetValue -> compareConstValue { it.compareIntegerNumericConst(compare) }
```

Comment:
The changes to check for `val`s are good! Could you please expand on the tests too?